### ISSUE TITLE
Nested panels not rendering correctly

### ISSUE THREAD

**Issue by adrhill** [FedeClaudi/Term.jl#167]
Rendering nested panels [as shown in the docs](https://fedeclaudi.github.io/Term.jl/stable/layout/nesting/#Nesting) doesn't work by default, requiring manual sizing on all inner panels.

In the following examples, the width of the inner panels should be inferred from the outer panel:
```Julia
julia> Panel(Panel("Foo"), Panel("Bar"))
╭──────────────────────────────────────────────────────────────────────────────────────╮
│  ╭────────────────────────────────────────────────────────────────────────────       │
│  ──────────╮                                                                         │
│  │                                                                                   │
│  Foo                                                                                 │
│     │                                                                                │
│  ╰────────────────────────────────────────────────────────────────────────────       │
│  ──────────╯                                                                         │
│  ╭────────────────────────────────────────────────────────────────────────────       │
│  ──────────╮                                                                         │
│  │                                                                                   │
│  Bar                                                                                 │
│     │                                                                                │
│  ╰────────────────────────────────────────────────────────────────────────────       │
│  ──────────╯                                                                         │
╰──────────────────────────────────────────────────────────────────────────────────────╯


julia> Panel(Panel("Foo"), Panel("Bar"); width=30)
╭────────────────────────────╮
│  ╭──────────────────       │
│  ───────────────────────   │
│  ───────────────────────   │
│  ──────────────────────╮│
│  [22m                      │
│  │                         │
│  Foo                       │
│                            │
│                            │
│                 │          │
│  ╰──────────────────       │
│  ───────────────────────   │
│  ───────────────────────   │
│  ──────────────────────╯│
│  [22m                      │
│  ╭──────────────────       │
│  ───────────────────────   │
│  ───────────────────────   │
│  ──────────────────────╮│
│  [22m                      │
│  │                         │
│  Bar                       │
│                            │
│                            │
│                 │          │
│  ╰──────────────────       │
│  ───────────────────────   │
│  ───────────────────────   │
│  ──────────────────────╯│
│  [22m                      │
╰────────────────────────────╯
```

Adding `width` for one inner panel gets rid of one of the `[22m` characters, but still fails:
```Julia
julia> Panel(Panel("Foo"), Panel("Bar"; width=20); width=30)
╭────────────────────────────╮
│  ╭──────────────────       │
│  ───────────────────────   │
│  ───────────────────────   │
│  ──────────────────────╮│
│  [22m                      │
│  │                         │
│  Foo                       │
│                            │
│                            │
│                 │          │
│  ╰──────────────────       │
│  ───────────────────────   │
│  ───────────────────────   │
│  ──────────────────────╯│
│  [22m                      │
│  ╭──────────────────       │
│  ╮                         │
│                            │
│                            │
│                            │
│  │  Bar                    │
│  │                         │
│                            │
│                            │
│                            │
│  ╰──────────────────       │
│  ╯                         │
│                            │
│                            │
│                            │
╰────────────────────────────╯
```
Specifying a width for all inner panels renders successfully:
```Julia
julia> Panel(Panel("Foo"; width=20), Panel("Bar"; width=20); width=30)
╭────────────────────────────╮
│  ╭──────────────────╮      │
│  │  Foo             │      │
│  ╰──────────────────╯      │
│  ╭──────────────────╮      │
│  │  Bar             │      │
│  ╰──────────────────╯      │
╰────────────────────────────╯


julia> Panel(Panel("Foo"; width=20), Panel("Bar"; width=20))
╭──────────────────────────────────────────────────────────────────────────────────────╮
│  ╭──────────────────╮                                                                │
│  │  Foo             │                                                                │
│  ╰──────────────────╯                                                                │
│  ╭──────────────────╮                                                                │
│  │  Bar             │                                                                │
│  ╰──────────────────╯                                                                │
╰──────────────────────────────────────────────────────────────────────────────────────╯
```

**Comment by FedeClaudi** [comment]
By default `Panel` has a fixed width. It's content gets broken up into multiple lines if its width exceeds the Panel's.
You can either: 1) decrease the width of the content, 2) increase the width of Panel or 3) use` fit=true`
<img width="492" alt="image" src="https://user-images.githubusercontent.com/17436313/203899710-49b4e9d6-12e9-4c7b-9745-ecc2758dfcad.png">

**Comment by adrhill** [comment]
This solution doesn't work for me on Term `v1.0.5` and iTerm Build `3.4.4`. 
According to a quick Google search, 80 is the "default" terminal width:
```Julia
julia> w = Term.console_width()
80

julia> dw = 6; # difference in width between nested panels

julia> Panel(Panel("Foo"), Panel("Bar"), fit=true)
╭──────────────────────────────────────────────────────────────────────────────╮
│  ╭────────────────────────────────────────────────────────────────────       │
│  ──────────╮                                                                 │
│  │                                                                           │
│  Foo                                                                         │
│     │                                                                        │
│  ╰────────────────────────────────────────────────────────────────────       │
│  ──────────╯                                                                 │
│  ╭────────────────────────────────────────────────────────────────────       │
│  ──────────╮                                                                 │
│  │                                                                           │
│  Bar                                                                         │
│     │                                                                        │
│  ╰────────────────────────────────────────────────────────────────────       │
│  ──────────╯                                                                 │
╰──────────────────────────────────────────────────────────────────────────────╯


julia> Panel(Panel("Foo", width=w-dw), Panel("Bar", width=w-dw), width=w) # workaround
╭──────────────────────────────────────────────────────────────────────────────╮
│  ╭────────────────────────────────────────────────────────────────────────╮  │
│  │  Foo                                                                   │  │
│  ╰────────────────────────────────────────────────────────────────────────╯  │
│  ╭────────────────────────────────────────────────────────────────────────╮  │
│  │  Bar                                                                   │  │
│  ╰────────────────────────────────────────────────────────────────────────╯  │
╰──────────────────────────────────────────────────────────────────────────────╯


julia> w = Term.console_width() # I resized my terminal here
50

julia> Panel(Panel("Foo"), Panel("Bar"), fit=true)

╭────────────────────────────────────────────────╮
│  ╭──────────────────────────────────────       │
│  ──────────╮                                   │
│  │                                             │
│  Foo                                           │
│     │                                          │
│  ╰──────────────────────────────────────       │
│  ──────────╯                                   │
│  ╭──────────────────────────────────────       │
│  ──────────╮                                   │
│  │                                             │
│  Bar                                           │
│     │                                          │
│  ╰──────────────────────────────────────       │
│  ──────────╯                                   │
╰────────────────────────────────────────────────╯


julia> Panel(Panel("Foo", width=w-dw), Panel("Bar", width=w-dw), width=w) # workaround
╭────────────────────────────────────────────────╮
│  ╭──────────────────────────────────────────╮  │
│  │  Foo                                     │  │
│  ╰──────────────────────────────────────────╯  │
│  ╭──────────────────────────────────────────╮  │
│  │  Bar                                     │  │
│  ╰──────────────────────────────────────────╯  │
╰────────────────────────────────────────────────╯

```

With more deeply nested Panels and dynamic content, it gets quite bothersome to pass the width around.

From my perspective as a user, a sensible default would be to use `width = console_width() - dw * depth` for all panels, the outermost panel having `depth=0`.

**Comment by FedeClaudi** [comment]
What's happening here is that `Panel` has a [default width](https://github.com/FedeClaudi/Term.jl/blob/f772f28c4b352f3e56de5d59476b72cfddbdd2f3/src/Term.jl#L38) which is `88` or the terminal width, whichever is smaller. 
With the terminal width at 80, the default width is 80. When you call the first `Panel("Foo")` that gets created with a width of 80. When you put it into the outer, that too has a width of 80 so the content won't fit anymore hence the layout breaking. 
Term doesn't have a way to dynamically create/alter renderables and it doesn't have a sense of nested hierarchy. 
So the `depth` information is just not there. Consider also that Panel could have any sort of different things inside.

So basically now the first call to `Panel("Foo")` actually creates the panel's text (border and content), it's not just a place-holder for something that gets created later hence why it doesn't respond to its context. Having a system that creates the text at render time and is more context aware would be useful, but it's beyond the current scope of Term's functionality.

**Comment by FedeClaudi** [comment]
That said, especially on narrow terminals, I can't imagine anyone going deeper than 3 nested layers because you then end up having very little space in each panel, so the keeping track of the widths is hopefully not too cumbersome.

**Comment by adrhill** [comment]
Thanks for the answer, I'll specify my widths manually. :)

I still run into errors with text reshaping when using nested panels:
```Julia
julia> loreipsum = "Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit amet.";

julia> w = console_width()
88

julia> dw = 6; # difference in width between nested panels

julia> Panel(Panel(loreipsum; width=w-dw); width=w) # works
╭──────────────────────────────────────────────────────────────────────────────────────╮
│  ╭────────────────────────────────────────────────────────────────────────────────╮  │
│  │  Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy      │  │
│  │  eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam      │  │
│  │  voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet       │  │
│  │  clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit       │  │
│  │  amet. Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam       │  │
│  │  nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed    │  │
│  │  diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum.       │  │
│  │  Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor      │  │
│  │  sit amet.                                                                     │  │
│  ╰────────────────────────────────────────────────────────────────────────────────╯  │
╰──────────────────────────────────────────────────────────────────────────────────────╯


julia> Panel(loreipsum, Panel(; width=w-dw); width=w) # fails
╭──────────────────────────────────────────────────────────────────────────────────────╮
│  Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod     │
│  tempor invidunt ut labore et dolore magna                                           │
│  aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et    │
│  ea rebum. Stet clita kasd gubergren, no sea                                         │
│  takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum dolor sit amet,        │
│  consetetur sadipscing elitr, sed diam nonumy                                        │
│  eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam            │
│  voluptua. At vero eos et accusam et justo duo                                       │
│  dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est         │
│  Lorem ipsum dolor sit amet.                                                         │
│  ╭────────────────────────────────────────────────────────────────────────────       │
│  ────╮                                                                               │
│  │                                                                                   │
│  │                                                                                   │
│  ╰────────────────────────────────────────────────────────────────────────────       │
│  ────╯                                                                               │
╰──────────────────────────────────────────────────────────────────────────────────────╯


julia> Panel(loreipsum, Panel(""; width=w-dw); width=w) # fails
╭──────────────────────────────────────────────────────────────────────────────────────╮
│  Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod     │
│  tempor invidunt ut labore et dolore magna                                           │
│  aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et    │
│  ea rebum. Stet clita kasd gubergren, no sea                                         │
│  takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum dolor sit amet,        │
│  consetetur sadipscing elitr, sed diam nonumy                                        │
│  eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam            │
│  voluptua. At vero eos et accusam et justo duo                                       │
│  dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est         │
│  Lorem ipsum dolor sit amet.                                                         │
│  ╭────────────────────────────────────────────────────────────────────────────       │
│  ────╮                                                                               │
│  │                                                                                   │
│  │                                                                                   │
│  ╰────────────────────────────────────────────────────────────────────────────       │
│  ────╯                                                                               │
╰──────────────────────────────────────────────────────────────────────────────────────╯


julia> Panel(loreipsum, Panel(loreipsum; width=w-dw); width=w) # fails
╭──────────────────────────────────────────────────────────────────────────────────────╮
│  Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod     │
│  tempor invidunt ut labore et dolore magna                                           │
│  aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et    │
│  ea rebum. Stet clita kasd gubergren, no sea                                         │
│  takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum dolor sit amet,        │
│  consetetur sadipscing elitr, sed diam nonumy                                        │
│  eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam            │
│  voluptua. At vero eos et accusam et justo duo                                       │
│  dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est         │
│  Lorem ipsum dolor sit amet.                                                         │
│  ╭────────────────────────────────────────────────────────────────────────────       │
│  ────╮                                                                               │
│  │  Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy         │
│  │                                                                                   │
│  │  eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam         │
│  │                                                                                   │
│  │  voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet          │
│  │                                                                                   │
│  │  clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit          │
│  │                                                                                   │
│  │  amet. Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam          │
│  │                                                                                   │
│  │  nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed       │
│  │                                                                                   │
│  │  diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum.          │
│  │                                                                                   │
│  │  Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor         │
│  │                                                                                   │
│  │  sit amet.                                                                        │
│  │                                                                                   │
│  ╰────────────────────────────────────────────────────────────────────────────       │
│  ────╯                                                                               │
╰──────────────────────────────────────────────────────────────────────────────────────╯


julia> Panel(loreipsum, Panel(loreipsum; width=w-dw, fit=false); width=w, fit=false) # fails
╭──────────────────────────────────────────────────────────────────────────────────────╮
│  Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod     │
│  tempor invidunt ut labore et dolore magna                                           │
│  aliquyam erat, sed diam voluptua. At vero eos et accusam et justo duo dolores et    │
│  ea rebum. Stet clita kasd gubergren, no sea                                         │
│  takimata sanctus est Lorem ipsum dolor sit amet. Lorem ipsum dolor sit amet,        │
│  consetetur sadipscing elitr, sed diam nonumy                                        │
│  eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam            │
│  voluptua. At vero eos et accusam et justo duo                                       │
│  dolores et ea rebum. Stet clita kasd gubergren, no sea takimata sanctus est         │
│  Lorem ipsum dolor sit amet.                                                         │
│  ╭────────────────────────────────────────────────────────────────────────────       │
│  ────╮                                                                               │
│  │  Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy         │
│  │                                                                                   │
│  │  eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed diam         │
│  │                                                                                   │
│  │  voluptua. At vero eos et accusam et justo duo dolores et ea rebum. Stet          │
│  │                                                                                   │
│  │  clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor sit          │
│  │                                                                                   │
│  │  amet. Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam          │
│  │                                                                                   │
│  │  nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam erat, sed       │
│  │                                                                                   │
│  │  diam voluptua. At vero eos et accusam et justo duo dolores et ea rebum.          │
│  │                                                                                   │
│  │  Stet clita kasd gubergren, no sea takimata sanctus est Lorem ipsum dolor         │
│  │                                                                                   │
│  │  sit amet.                                                                        │
│  │                                                                                   │
│  ╰────────────────────────────────────────────────────────────────────────────       │
│  ────╯                                                                               │
╰──────────────────────────────────────────────────────────────────────────────────────╯


```

**Comment by FedeClaudi** [comment]
mmm yes, that might be a bug. Text wrapping has been a pain since day one, and it's clearly not perfect still. 
I'm working on a `macro` to automate nesting, maybe that'll help.

**Comment by FedeClaudi** [comment]
I think you'll like this:
```julia
using Term



@nested_panels Panel(
    Panel(
        Panel(
            Panel(
                ; title="deepest", style="blue", background="blue"
            ); title="deeper", style="green");
        title="not so deep");
    style="red",
    title="outer"
) 

```


<img width="531" alt="image" src="https://user-images.githubusercontent.com/17436313/204070212-f1835e83-57a6-4306-a15c-df006ab383e7.png">
<img width="806" alt="image" src="https://user-images.githubusercontent.com/17436313/204070217-33ae9732-45c9-4495-babd-12d4d30655a2.png">

It sets the width of each panel to match their depth. It only works if you're actively calling `Panel()` in the macro call. 
something like this won't work:
```julia
another_panel = Panel("this won't work")
@nested_panels Panel(
    Panel(
        Panel(
            Panel(
                another_panel,
                ; title="deepest", 
            ); title="deeper", style="green");
        title="not so deep");
    style="red",
    title="outer"
) 


```

**Comment by adrhill** [comment]
This looks great, thanks a lot! :smiley:

**Comment by FedeClaudi** [comment]
The macro is available with Term 1.0.6 which I just released. Let me know if you come upon any bug using it!

**Comment by adrhill** [comment]
This works really great for my use-cases, thanks a lot!

Trying to break the macro, I found an edge-case where Panels that wrap other Panels will stop the contents of the inner Panel from rendering:
```Julia
julia> @nested_panels Panel("Foo") # works
╭──────────────────────────────────────────────────────────────────────────────────────────╮
│  Foo                                                                                     │
╰──────────────────────────────────────────────────────────────────────────────────────────╯


julia> @nested_panels Panel("Foo", Panel("Bar")) # works
╭──────────────────────────────────────────────────────────────────────────────────────────╮
│  Foo                                                                                     │
│  ╭────────────────────────────────────────────────────────────────────────────────────╮  │
│  │  Bar                                                                               │  │
│  ╰────────────────────────────────────────────────────────────────────────────────────╯  │
╰──────────────────────────────────────────────────────────────────────────────────────────╯


julia> @nested_panels Panel(Panel("Bar")) # empty
╭──────────────────────────────────────────────────────────────────────────────────────────╮
│                                                                                          │
╰──────────────────────────────────────────────────────────────────────────────────────────╯


julia> @nested_panels Panel(Panel("Bar"), "Qux") # empty
╭──────────────────────────────────────────────────────────────────────────────────────────╮
│                                                                                          │
╰──────────────────────────────────────────────────────────────────────────────────────────╯


```

**Comment by FedeClaudi** [comment]
Happy to hear that you're happy with it. 

I've discovered the bug right after releasing (ofc), but it's fixed by a1b52810cc08f1484736f0f41c12bbf4d9b81b49
```

julia> @nested_panels Panel(Panel("Bar"), "Qux") # empty
╭──────────────────────────────────────────────────────────────────────────────────────╮
│  ╭────────────────────────────────────────────────────────────────────────────────╮  │
│  │  Bar                                                                           │  │
│  ╰────────────────────────────────────────────────────────────────────────────────╯  │
│  Qux                                                                                 │
╰──────────────────────────────────────────────────────────────────────────────────────╯
```

It will be patched in the next release. In the meanwhile, adding a keyword argument (e.g. `; style="default"`) will circumvent the problem.

**Comment by adrhill** [comment]
Thanks! :)
